domain/removal/state/model/operation.go

```diff
@@ -0,0 +1,249 @@
+	}
+
+	var result []task
+	if err := tx.Query(ctx, stmt, machine{UUID: machineUUID}).GetAll(&result); err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+		return nil, errors.Errorf("running task UUIDs query: %w", err)
+	}
+
+	return transform.Slice(result, func(t task) string {
+		return t.UUID
+	}), nil
+}
+
+// getOperationByTaskUUIDs returns the UUIDs of operations that contains the tasks.
+func (st *State) getOperationByTaskUUIDs(ctx context.Context, tx *sqlair.TX, taskUUIDs []string) ([]string, error) {
+	if len(taskUUIDs) == 0 {
+		return nil, nil
+	}
+
+	type operation entityUUID
+	type tasks uuids
+
+	stmt, err := st.Prepare(`
+SELECT DISTINCT operation_uuid AS &operation.uuid
+FROM   operation_task
+WHERE  uuid IN ($tasks[:])`, operation{}, tasks{})
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	var result []operation
+	if err := tx.Query(ctx, stmt, tasks(taskUUIDs)).GetAll(&result); err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+		return nil, errors.Errorf("running operation UUIDs query: %w", err)
+	}
+
+	return transform.Slice(result, func(op operation) string {
+		return op.UUID
+	}), nil
+}
+
+// getEmptyOperationUUIDs returns the UUIDs of operations that are empty.
+func (st *State) getEmptyOperationUUIDs(ctx context.Context, tx *sqlair.TX, operationUUIDs []string) ([]string, error) {
+	if len(operationUUIDs) == 0 {
+		return nil, nil
+	}
+	type operation entityUUID
+
+	stmt, err := st.Prepare(`
+SELECT DISTINCT o.uuid AS &operation.uuid
+FROM operation AS o
+LEFT JOIN operation_task AS ot ON o.uuid = ot.operation_uuid 
+WHERE o.uuid IN ($uuids[:])
+AND ot.uuid IS NULL`, operation{}, uuids{})
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	var result []operation
+	if err := tx.Query(ctx, stmt, uuids(operationUUIDs)).GetAll(&result); err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+		return nil, errors.Errorf("running get empty operation query: %w", err)
+	}
+
+	return transform.Slice(result, func(op operation) string {
+		return op.UUID
+	}), nil
+}
+
+// deleteTaskByUUIDs deletes the tasks and their associated dependencies by
+// their UUIDs.
+// It returns the path of object_store_metadata_path to allow cleanup
+// of the object datastore in the service layer.
+func (st *State) deleteTaskByUUIDs(ctx context.Context, tx *sqlair.TX, toDelete []string) ([]string, error) {
+	if len(toDelete) == 0 {
+		return nil, nil
+	}
+
+	// get store paths
+	type store struct {
+		Path string `db:"path"`
+	}
+	tasks := uuids(toDelete)
+	stmt, err := st.Prepare(`
+SELECT &store.path
+FROM object_store_metadata_path AS osp
+JOIN operation_task_output AS oto ON osp.metadata_uuid = oto.store_uuid
+WHERE task_uuid IN ($uuids[:])`, tasks, store{})
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+	var stores []store
+	if err := tx.Query(ctx, stmt, tasks).GetAll(&stores); err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+		return nil, errors.Errorf("getting task store uuids: %w", err)
+	}
```

Comment:
nit: this could be done as a separate method, since the root method is now doing two (quite) distinct things.